[LibOS] procfs: Add support for /proc/self/cmdline

svenkata9 opened the pull request:
Signed-off-by: Sankaranarayanan Venkatasubramanian <svenkat9>

<!--
    Please fill in the following form before submitting this PR
    and ensure that your code follows our coding style guideline:
    https://graphene.readthedocs.io/en/latest/devel/coding-style.html -->

## Description of the changes <!-- (reasons and measures) -->

<!--
    If your PR fixes an issue, please remember to add "Fixes #issue_number"
    here, to automatically close it on merge. -->

## How to test this PR? <!-- (if applicable) -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/oscarlab/graphene/2180)
<!-- Reviewable:end -->

dimakuv reviewed (changes requested):
> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: all files reviewed, 14 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @svenkata9)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU81riR5-h54OIIQUXa:-MU81riR5-h54OIIQUXb:bpn2c4j) (no related file):*
@svenkata9 There are two `signed-off-by` lines in the commit description. Please remove the one without the email.

---
___
*[LibOS/shim/include/shim_process.h, line 62 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU825vA5iQmmGKNLBO9:-MU825vA5iQmmGKNLBOA:bl7hh15) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/include/shim_process.h#L62)):*
> ```objc
>     struct shim_lock fs_lock;
> 
>     /* cmdline */
> ```

Could you expand this comment? Like this:
```
/* complete command line for the process, as reported by /proc/[pid]/cmdline */
```
___
*[LibOS/shim/include/shim_process.h, line 64 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU820uX1DXG7tgnKMbd:-MU820uX1DXG7tgnKMbe:bx3g4u2) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/include/shim_process.h#L64)):*
> ```objc
>     /* cmdline */
>     char cmdline[STR_SIZE];
>     unsigned int cmdline_length;
> ```

`unsigned int` -> `size_t`
___
*[LibOS/shim/src/bookkeep/shim_process.c, line 65 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU82UDW4kLNGiYce2R5:-MU82UDW4kLNGiYce2R6:bprsyzf) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/bookkeep/shim_process.c#L65)):*
> ```C
>     /* The command line arguments passed are stored in /proc/self/cmdline as
>      * part of the proc fs. They are not separated by space, but by NUL instead.
>      * So, it is essential to maintain the cmdline_length also as a member here */
> ```

We have a 100-char wrap limit in C files, so please re-adjust these lines to 100 chars.

Please also add a dot at the end of this comment: `...as a member here. */`
___
*[LibOS/shim/src/bookkeep/shim_process.c, line 68 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU85d0w63TQTBuJA4Wu:-MU85d0w63TQTBuJA4Wv:b-t5sz4o) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/bookkeep/shim_process.c#L68)):*
> ```C
> 
>     g_process.cmdline_length = 0;
>     memset(g_process.cmdline, '\0', STR_SIZE);
> ```

`STR_SIZE` -> `ARRAY_SIZE(g_process.cmdline)`
___
*[LibOS/shim/src/bookkeep/shim_process.c, line 69 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU82sVx8qhN-CsCSGbu:-MU82sVy8KXxb3NmpQsv:bhmjw3p) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/bookkeep/shim_process.c#L69)):*
> ```C
>     g_process.cmdline_length = 0;
>     memset(g_process.cmdline, '\0', STR_SIZE);
>     int tmplen = 0;
> ```

`int` -> `size_t`
___
*[LibOS/shim/src/bookkeep/shim_process.c, line 72 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU84tE79pC970jfKNbN:-MU84tE8-UYSYKz61hPz:bdo4w81) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/bookkeep/shim_process.c#L72)):*
> ```C
> 
>     for (int i = 0; i < argc; i++) {
>         memcpy(g_process.cmdline + tmplen, argv[i], strlen(argv[i]));
> ```

Before doing all this, we need to check if there is still available space in the `cmdline` buffer. Smth like this:
```
if (tmplen + strlen(argv[i]) + 1 > ARRAY_SIZE(g_process.cmdline))
    return -ENOMEM;
```
___
*[LibOS/shim/src/bookkeep/shim_process.c, line 77 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU83b5-0EEiL478NXH_:-MU83b5-0EEiL478NXHa:b-ksmrbd) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/bookkeep/shim_process.c#L77)):*
> ```C
>     }
> 
>     g_process.cmdline_length = --tmplen;
> ```

Why is there `--`? Looks like a bug. (For example, imagine if `argc == 0`, then `cmdline_length == -1 ???`).
___
*[LibOS/shim/src/bookkeep/shim_process.c, line 228 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU83sT9322Lp7uWwcVV:-MU83sT9322Lp7uWwcVW:b-jyofc7) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/bookkeep/shim_process.c#L228)):*
> ```C
>     new_process->pgid = process->pgid;
> 
>     /* copy cmdline parameters in case of a fork */
> ```

The comment message is a bit useless (we know that we're doing a fork/clone if we are in this checkpointing function, and we also see that we copy `cmdline` parameters from the source code).

However, I think it's important to highlight the trick we play with copying from `g_process`. So maybe replace your comment with this:
```
/* copy cmdline (used by /proc/[pid]/cmdline) from the current process */
```
___
*[LibOS/shim/src/bookkeep/shim_process.c, line 229 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU85iTm1NPrYz7bb2Ro:-MU85iTm1NPrYz7bb2Rp:b-t5sz4o) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/bookkeep/shim_process.c#L229)):*
> ```C
> 
>     /* copy cmdline parameters in case of a fork */
>     memcpy(new_process->cmdline, g_process.cmdline, STR_SIZE);
> ```

`STR_SIZE` -> `ARRAY_SIZE(g_process.cmdline)`
___
*[LibOS/shim/src/fs/proc/thread.c, line 610 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU865xpDVIppFCqZhnN:-MU865xpDVIppFCqZhnO:bhmjw3p) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/fs/proc/thread.c#L610)):*
> ```C
>         return ret;
> 
>     int buffer_size = g_process.cmdline_length;
> ```

`int` -> `size_t`
___
*[LibOS/shim/src/fs/proc/thread.c, line 615 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU86HMT1R8GjscUnA8L:-MU86HMU52hGxWHzbBEu:by6qf7p) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/fs/proc/thread.c#L615)):*
> ```C
>         ret = -ENOMEM;
>         return ret;
>     }
> ```

A redundant line. Simply:
```
    if (!buffer)
        return -ENOMEM;
```
___
*[LibOS/shim/src/fs/proc/thread.c, line 619 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU86QUq3Mn9xymN75wE:-MU86QUq3Mn9xymN75wF:b-b4054j) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/fs/proc/thread.c#L619)):*
> ```C
>     memcpy(buffer, g_process.cmdline, buffer_size);
> 
>     struct shim_str_data* data = calloc(1, sizeof(struct shim_str_data));
> ```

`sizeof(struct shim_str_data)` -> `sizeof(*data)`
___
*[LibOS/shim/src/fs/proc/thread.c, line 623 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU86U_UC6M17NsnPiJ7:-MU86U_UC6M17NsnPiJ8:b-plug9k) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/fs/proc/thread.c#L623)):*
> ```C
>         free(buffer);
>         ret = -ENOMEM;
>         return ret;
> ```

A redundant line. Simply: `return -ENOMEM;`


<!-- Sent from Reviewable.io -->

mkow reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: all files reviewed, 15 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @svenkata9)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU8phb-0AR6x70oNnC5:-MU8phb-0AR6x70oNnC6:b-oux3sl) (no related file):*
Please fix the email address in your commit so that GitHub can recognize it (either by changing it in the commit or by adding it to your GitHub account). You can push-force if you choose the former.

Also, the component tag in the one-liner should be `[LibOS]`, you can do `[LibOS] procfs:` if you want to mention the sub-component also.

---


<!-- Sent from Reviewable.io -->

svenkata9 commented:
Thanks for the comments. Updated the PR with changes.

mkow reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: 1 of 4 files reviewed, 15 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv and @svenkata9)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU8phb-0AR6x70oNnC5:-MU91sWZ8na0bo5dxyWw:b-gukbb) (no related file):*
> Thanks for the comments. Updated the PR with changes.

@svenkata9 This is still not fixed :) Btw. it also should be "Add", not "Adding". Please read our contributing guide ;)

---


<!-- Sent from Reviewable.io -->

svenkata9 commented:
> > _[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)_ status: 1 of 4 files reviewed, 15 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv and @svenkata9)
> 
> _[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU8phb-0AR6x70oNnC5:-MU91sWZ8na0bo5dxyWw:b-gukbb) (no related file):_
> 
> > Thanks for the comments. Updated the PR with changes.
> 
> @svenkata9 This is still not fixed :) Btw. it also should be "Add", not "Adding". Please read our contributing guide ;)

Missed that comment. Fixed it.

mkow reviewed (changes requested):
> Reviewed 1 of 4 files at r1, 1 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: 2 of 4 files reviewed, 20 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv and @svenkata9)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU8phb-0AR6x70oNnC5:-MU9KTkz9gY9L7HZvwPr:b-6u3y0d) (no related file):*
<details><summary><i>Previously, svenkata9 wrote…</i></summary><blockquote>

Missed that comment. Fixed it. 
</blockquote></details>

Thanks.
btw. No need to click "request review" after each change, once someone commented under a PR then they are automatically added to the watch list and receive emails ;)

---
___
*[LibOS/shim/include/shim_process.h, line 63 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBG2JTDmt2083YHtzU:-MUBG2JTDmt2083YHtzV:b-uy05zr) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/include/shim_process.h#L63)):*
> ```objc
> 
>     /* complete command line for the process, as reported by /proc/[pid]/cmdline */
>     char cmdline[STR_SIZE];
> ```

`STR_SIZE` is 4096 currently, I think cmdlines on Linux can be much longer than this. Any reason to hardcode the size instead of allocating it dynamically?
___
*[LibOS/shim/include/shim_process.h, line 64 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBGJ9N6vnbVvw9l2pa:-MUBGJ9N6vnbVvw9l2pb:b-srks1j) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/include/shim_process.h#L64)):*
> ```objc
>     /* complete command line for the process, as reported by /proc/[pid]/cmdline */
>     char cmdline[STR_SIZE];
>     size_t cmdline_length;
> ```

is this commandline length or size?
___
*[LibOS/shim/src/fs/proc/thread.c, line 611 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBIAfNEdS3obqDD3rT:-MUBIAfNEdS3obqDD3rU:bh83t4z) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L611)):*
> ```C
> 
>     size_t buffer_size = g_process.cmdline_length;
>     buffer = calloc(buffer_size, sizeof(char));
> ```

Why not `malloc`? You overwrite it right below.
Or even better, `strdup()`.
___
*[LibOS/shim/src/fs/proc/thread.c, line 618 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBJIo32jNDO5MR45TB:-MUBJIo32jNDO5MR45TC:bbcu709) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L618)):*
> ```C
>     memcpy(buffer, g_process.cmdline, buffer_size);
> 
>     struct shim_str_data* data = calloc(1, sizeof(*data));
> ```

ditto, why not malloc?
___
*[LibOS/shim/src/fs/proc/thread.c, line 632 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBI2g-2x8GH257ZoUt:-MUBI2g0802mSJtLm7_7:b-l0o1r3) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L632)):*
> ```C
> 
>     // Don't really need to do any work here, but keeping as a placeholder,
>     // just in case.
> ```

I don't understand this comment, what placeholder?
___
*[LibOS/shim/src/fs/proc/thread.c, line 647 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBJVo7CBaON18kM7yL:-MUBJVo7CBaON18kM7yM:b-d4ui68) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L647)):*
> ```C
>     // Only used by one file
>     __UNUSED(name);
>     memset(buf, 0, sizeof(struct stat));
> ```

`sizeof(struct stat)` -> `sizeof(*buf)`


<!-- Sent from Reviewable.io -->

svenkata9 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: 1 of 4 files reviewed, 20 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv, @mkow, and @svenkata9)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU81riR5-h54OIIQUXa:-MUBXx3D7mP34m-KN4Eq:b-896fix) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

@svenkata9 There are two `signed-off-by` lines in the commit description. Please remove the one without the email.
</blockquote></details>

Done.

---
___
*[LibOS/shim/include/shim_process.h, line 62 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU825vA5iQmmGKNLBO9:-MUBY01u97dKJ6Lnjhi8:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/include/shim_process.h#L62)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Could you expand this comment? Like this:
```
/* complete command line for the process, as reported by /proc/[pid]/cmdline */
```
</blockquote></details>

Done.
___
*[LibOS/shim/include/shim_process.h, line 64 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU820uX1DXG7tgnKMbd:-MUBY-dg4Rv4H8uSYwQM:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/include/shim_process.h#L64)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`unsigned int` -> `size_t`
</blockquote></details>

Done.
___
*[LibOS/shim/include/shim_process.h, line 63 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBG2JTDmt2083YHtzU:-MUBY0Pp0EERGlCXHmPw:b-7cx0i6) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/include/shim_process.h#L63)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

`STR_SIZE` is 4096 currently, I think cmdlines on Linux can be much longer than this. Any reason to hardcode the size instead of allocating it dynamically?
</blockquote></details>

Discussed with @dimakuv. As I read elsewhere in the web, it can be as large as 256KB (there is no definitive information on this though). There is no real reason to keep it at 4096 here except that it appears to be a decent size and large enough to handle the current needs.
___
*[LibOS/shim/include/shim_process.h, line 64 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBGJ9N6vnbVvw9l2pa:-MUBYly06_SiDRuhute8:b-o0w6ad) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/include/shim_process.h#L64)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

is this commandline length or size?
</blockquote></details>

length
___
*[LibOS/shim/src/bookkeep/shim_process.c, line 65 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU82UDW4kLNGiYce2R5:-MUBYofC3F9Z56iQlDeh:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/bookkeep/shim_process.c#L65)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

We have a 100-char wrap limit in C files, so please re-adjust these lines to 100 chars.

Please also add a dot at the end of this comment: `...as a member here. */`
</blockquote></details>

Done.
___
*[LibOS/shim/src/bookkeep/shim_process.c, line 68 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU85d0w63TQTBuJA4Wu:-MUBYrX_9IelV4nRHXk-:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/bookkeep/shim_process.c#L68)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`STR_SIZE` -> `ARRAY_SIZE(g_process.cmdline)`
</blockquote></details>

Done.
___
*[LibOS/shim/src/bookkeep/shim_process.c, line 69 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU82sVx8qhN-CsCSGbu:-MUBYpW4BTucYBIUyMsf:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/bookkeep/shim_process.c#L69)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`int` -> `size_t`
</blockquote></details>

Done.
___
*[LibOS/shim/src/bookkeep/shim_process.c, line 72 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU84tE79pC970jfKNbN:-MUBYroSAE-7SECw21NX:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/bookkeep/shim_process.c#L72)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Before doing all this, we need to check if there is still available space in the `cmdline` buffer. Smth like this:
```
if (tmplen + strlen(argv[i]) + 1 > ARRAY_SIZE(g_process.cmdline))
    return -ENOMEM;
```
</blockquote></details>

Done.
___
*[LibOS/shim/src/bookkeep/shim_process.c, line 77 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU83b5-0EEiL478NXH_:-MUBYqdC352eDB27qRrA:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/bookkeep/shim_process.c#L77)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Why is there `--`? Looks like a bug. (For example, imagine if `argc == 0`, then `cmdline_length == -1 ???`).
</blockquote></details>

Done.
___
*[LibOS/shim/src/bookkeep/shim_process.c, line 228 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU83sT9322Lp7uWwcVV:-MUBYr5zDryixybGJW67:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/bookkeep/shim_process.c#L228)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

The comment message is a bit useless (we know that we're doing a fork/clone if we are in this checkpointing function, and we also see that we copy `cmdline` parameters from the source code).

However, I think it's important to highlight the trick we play with copying from `g_process`. So maybe replace your comment with this:
```
/* copy cmdline (used by /proc/[pid]/cmdline) from the current process */
```
</blockquote></details>

Done.
___
*[LibOS/shim/src/bookkeep/shim_process.c, line 229 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU85iTm1NPrYz7bb2Ro:-MUBYsK54-KVbl5c0tI7:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/bookkeep/shim_process.c#L229)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`STR_SIZE` -> `ARRAY_SIZE(g_process.cmdline)`
</blockquote></details>

Done.
___
*[LibOS/shim/src/fs/proc/thread.c, line 610 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU865xpDVIppFCqZhnN:-MUBYv3S3GOOTvqsYJe_:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/fs/proc/thread.c#L610)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`int` -> `size_t`
</blockquote></details>

Done.
___
*[LibOS/shim/src/fs/proc/thread.c, line 615 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU86HMT1R8GjscUnA8L:-MUBZUeHC2HsW6CNFtoT:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/fs/proc/thread.c#L615)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

A redundant line. Simply:
```
    if (!buffer)
        return -ENOMEM;
```
</blockquote></details>

Done.
___
*[LibOS/shim/src/fs/proc/thread.c, line 619 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU86QUq3Mn9xymN75wE:-MUBZheGAG0VKkC5vPmu:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/fs/proc/thread.c#L619)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`sizeof(struct shim_str_data)` -> `sizeof(*data)`
</blockquote></details>

Done.
___
*[LibOS/shim/src/fs/proc/thread.c, line 623 at r1](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU86U_UC6M17NsnPiJ7:-MUBZsae7C5KDNESqNPg:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/77005aa7fddddfcfc1679eb7416c5ce73c916f4e/LibOS/shim/src/fs/proc/thread.c#L623)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

A redundant line. Simply: `return -ENOMEM;`
</blockquote></details>

Done.
___
*[LibOS/shim/src/fs/proc/thread.c, line 618 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBJIo32jNDO5MR45TB:-MUBZkSg65saucXMLsIR:b-ix2y5n) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L618)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

ditto, why not malloc?
</blockquote></details>

Following the existing practice in ..._maps_open...
___
*[LibOS/shim/src/fs/proc/thread.c, line 647 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBJVo7CBaON18kM7yL:-MUBZxP73Gq_QnI-BaBt:b9vgjic) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L647)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

`sizeof(struct stat)` -> `sizeof(*buf)`
</blockquote></details>

Present in the existing code, so following that....


<!-- Sent from Reviewable.io -->

svenkata9 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: 1 of 4 files reviewed, 20 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv and @mkow)

___
*[LibOS/shim/src/fs/proc/thread.c, line 611 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBIAfNEdS3obqDD3rT:-MUC5Z2O-tPJlAcQwHo3:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L611)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Why not `malloc`? You overwrite it right below.
Or even better, `strdup()`.
</blockquote></details>

Done.
___
*[LibOS/shim/src/fs/proc/thread.c, line 632 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBI2g-2x8GH257ZoUt:-MUC5UWuCgnj20pCs-sw:bvpeo2n) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L632)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

I don't understand this comment, what placeholder?
</blockquote></details>

Removed. Done.


<!-- Sent from Reviewable.io -->

dimakuv reviewed (changes requested):
> Reviewed 1 of 3 files at r2, 2 of 2 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: all files reviewed, 8 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @mkow and @svenkata9)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUCuWZl2KFl-WpUpMIo:-MUCuWZl2KFl-WpUpMIp:bn7bugi) (no related file):*
For future: please don't use `git merge`. We are working in the `git rebase` flow: https://www.atlassian.com/git/tutorials/merging-vs-rebasing. Note that you should never do `git rebase` by yourself; only the maintainers do `git rebase` during the final step of reviewing. Also check our guide: https://graphene.readthedocs.io/en/latest/devel/contributing.html#pull-requests

---
___
*[LibOS/shim/include/shim_process.h, line 63 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBG2JTDmt2083YHtzU:-MUCvn6q22wQYyKIqbBn:b-l1mcmw) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/include/shim_process.h#L63)):*
<details><summary><i>Previously, svenkata9 wrote…</i></summary><blockquote>

Discussed with @dimakuv. As I read elsewhere in the web, it can be as large as 256KB (there is no definitive information on this though). There is no real reason to keep it at 4096 here except that it appears to be a decent size and large enough to handle the current needs.
</blockquote></details>

Yes, we check whether the command-line args fit in 4096 anyway in `init_process()` and loudly fail if they don't. So we are safe in this sense.

We don't want to allocate dynamically because then we'll need more complicated code in checkpoint-restore.
___
*[LibOS/shim/include/shim_process.h, line 64 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBGJ9N6vnbVvw9l2pa:-MUCzcLT-YC1TMipnmxW:b-nyov93) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/include/shim_process.h#L64)):*
<details><summary><i>Previously, svenkata9 wrote…</i></summary><blockquote>

length
</blockquote></details>

I think it is a size. The difference is like this:
- "Length" is without the last NUL byte
- "Size" is with the last NUL byte

You have the last NUL byte in your `cmdline` buffer, so this variable must be renamed to `cmdline_size`. Sorry I haven't noticed this before.
___
*[LibOS/shim/src/bookkeep/shim_process.c, line 69 at r4](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUD19A_4P5SFqVb2bAk:-MUD19A_4P5SFqVb2bAl:banmrg7) ([raw file](https://github.com/oscarlab/graphene/blob/a5f328d87dd6b0bdc58330f508c776e57fc44230/LibOS/shim/src/bookkeep/shim_process.c#L69)):*
> ```C
>     g_process.cmdline_length = 0;
>     memset(g_process.cmdline, '\0', ARRAY_SIZE(g_process.cmdline));
>     size_t tmplen = 0;
> ```

Since we rename `cmdline_length` to `cmdline_size`, we should also rename `tmplen` -> `tmp_size`.
___
*[LibOS/shim/src/fs/proc/thread.c, line 618 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBJIo32jNDO5MR45TB:-MUD1g1lBflvm1wBlm5Y:b-ukr79q) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L618)):*
<details><summary><i>Previously, svenkata9 wrote…</i></summary><blockquote>

Following the existing practice in ..._maps_open...
</blockquote></details>

Yes, I would keep all these as-is, because they were copied from other places. So, we'll just change all these in one sweep in a future PR.
___
*[LibOS/shim/src/fs/proc/thread.c, line 632 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBI2g-2x8GH257ZoUt:-MUD23EO-pm7yvPu93RO:bvdnuxl) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L632)):*
<details><summary><i>Previously, svenkata9 wrote…</i></summary><blockquote>

Removed. Done.
</blockquote></details>

Again, this was taken from the other places. Our FS subsystem is full of these.


<!-- Sent from Reviewable.io -->

dimakuv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: all files reviewed, 8 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @mkow and @svenkata9)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUCuWZl2KFl-WpUpMIo:-MUD5UCX2urh_fkiGKoj:bllbppm) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

For future: please don't use `git merge`. We are working in the `git rebase` flow: https://www.atlassian.com/git/tutorials/merging-vs-rebasing. Note that you should never do `git rebase` by yourself; only the maintainers do `git rebase` during the final step of reviewing. Also check our guide: https://graphene.readthedocs.io/en/latest/devel/contributing.html#pull-requests
</blockquote></details>

I just noticed that you're using the `master` branch in your forked Graphene repo. That's why you need to do `git merge`... We require to create pull requests in separate, uniquely named branches, e.g. `borys/rework_pal_api` (see https://github.com/oscarlab/graphene/tree/borys/rework_pal_api and the corresponding PR https://github.com/oscarlab/graphene/pull/2182).

Well, it's fine for this PR, proceed with your `master` branch. But for the future, you'll have to create separate git branch for each new PR.

---


<!-- Sent from Reviewable.io -->

svenkata9 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: 1 of 4 files reviewed, 8 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv, @mkow, and @svenkata9)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUCuWZl2KFl-WpUpMIo:-MUD9YQ057bqWI5d9QFZ:bzeyi24) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I just noticed that you're using the `master` branch in your forked Graphene repo. That's why you need to do `git merge`... We require to create pull requests in separate, uniquely named branches, e.g. `borys/rework_pal_api` (see https://github.com/oscarlab/graphene/tree/borys/rework_pal_api and the corresponding PR https://github.com/oscarlab/graphene/pull/2182).

Well, it's fine for this PR, proceed with your `master` branch. But for the future, you'll have to create separate git branch for each new PR.
</blockquote></details>

Sure. 

---
___
*[LibOS/shim/include/shim_process.h, line 63 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBG2JTDmt2083YHtzU:-MUDDHSlDtJnvYx6_MdQ:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/include/shim_process.h#L63)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Yes, we check whether the command-line args fit in 4096 anyway in `init_process()` and loudly fail if they don't. So we are safe in this sense.

We don't want to allocate dynamically because then we'll need more complicated code in checkpoint-restore.
</blockquote></details>

Done.
___
*[LibOS/shim/include/shim_process.h, line 64 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBGJ9N6vnbVvw9l2pa:-MUDDE4e5-ZPBZMsh_Yr:b-bu91xm) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/include/shim_process.h#L64)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I think it is a size. The difference is like this:
- "Length" is without the last NUL byte
- "Size" is with the last NUL byte

You have the last NUL byte in your `cmdline` buffer, so this variable must be renamed to `cmdline_size`. Sorry I haven't noticed this before.
</blockquote></details>

Updated
___
*[LibOS/shim/src/fs/proc/thread.c, line 618 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBJIo32jNDO5MR45TB:-MUDD9yt6b4VjC8jIlEf:bb6un1j) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L618)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Yes, I would keep all these as-is, because they were copied from other places. So, we'll just change all these in one sweep in a future PR.
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

dimakuv reviewed (changes requested):
> Reviewed 3 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: all files reviewed, 7 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @mkow and @svenkata9)

___
*[LibOS/shim/src/bookkeep/shim_process.c, line 69 at r4](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUD19A_4P5SFqVb2bAk:-MUDVBkH6N5HtStjWmcs:b8h5ypa) ([raw file](https://github.com/oscarlab/graphene/blob/a5f328d87dd6b0bdc58330f508c776e57fc44230/LibOS/shim/src/bookkeep/shim_process.c#L69)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Since we rename `cmdline_length` to `cmdline_size`, we should also rename `tmplen` -> `tmp_size`.
</blockquote></details>

This is still not done.


<!-- Sent from Reviewable.io -->

svenkata9 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: 3 of 4 files reviewed, 7 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv and @mkow)

___
*[LibOS/shim/src/bookkeep/shim_process.c, line 69 at r4](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUD19A_4P5SFqVb2bAk:-MUDsIsw85EA-bjCD9uQ:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/a5f328d87dd6b0bdc58330f508c776e57fc44230/LibOS/shim/src/bookkeep/shim_process.c#L69)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

This is still not done.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

svenkata9 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: 3 of 4 files reviewed, 7 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv and @mkow)

___
*[LibOS/shim/src/fs/proc/thread.c, line 647 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBJVo7CBaON18kM7yL:-MUDsNkZ8Uc8KmAeE2rj:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L647)):*
<details><summary><i>Previously, svenkata9 wrote…</i></summary><blockquote>

Present in the existing code, so following that....
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dimakuv reviewed (approved):
> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: all files reviewed, 6 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @mkow)



<!-- Sent from Reviewable.io -->

svenkata9 commented:
> > Reviewed 1 of 4 files at r1, 1 of 3 files at r2.
> > _[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)_ status: 2 of 4 files reviewed, 20 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv and @svenkata9)
> 
> _[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2180#-MU8phb-0AR6x70oNnC5:-MU9KTkz9gY9L7HZvwPr:b-6u3y0d) (no related file):_
> 
> Previously, svenkata9 wrote…
> Thanks.
> btw. No need to click "request review" after each change, once someone commented under a PR then they are automatically added to the watch list and receive emails ;)
> 
> _[LibOS/shim/include/shim_process.h, line 63 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBG2JTDmt2083YHtzU:-MUBG2JTDmt2083YHtzV:b-uy05zr) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/include/shim_process.h#L63)):_
> 
> > ```objc
> >     /* complete command line for the process, as reported by /proc/[pid]/cmdline */
> >     char cmdline[STR_SIZE];
> > ```
> 
> `STR_SIZE` is 4096 currently, I think cmdlines on Linux can be much longer than this. Any reason to hardcode the size instead of allocating it dynamically?
> 
> _[LibOS/shim/include/shim_process.h, line 64 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBGJ9N6vnbVvw9l2pa:-MUBGJ9N6vnbVvw9l2pb:b-srks1j) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/include/shim_process.h#L64)):_
> 
> > ```objc
> >     /* complete command line for the process, as reported by /proc/[pid]/cmdline */
> >     char cmdline[STR_SIZE];
> >     size_t cmdline_length;
> > ```
> 
> is this commandline length or size?
> 
> _[LibOS/shim/src/fs/proc/thread.c, line 611 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBIAfNEdS3obqDD3rT:-MUBIAfNEdS3obqDD3rU:bh83t4z) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L611)):_
> 
> > ```c
> >     size_t buffer_size = g_process.cmdline_length;
> >     buffer = calloc(buffer_size, sizeof(char));
> > ```
> 
> Why not `malloc`? You overwrite it right below.
> Or even better, `strdup()`.
> 
> _[LibOS/shim/src/fs/proc/thread.c, line 618 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBJIo32jNDO5MR45TB:-MUBJIo32jNDO5MR45TC:bbcu709) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L618)):_
> 
> > ```c
> >     memcpy(buffer, g_process.cmdline, buffer_size);
> > 
> >     struct shim_str_data* data = calloc(1, sizeof(*data));
> > ```
> 
> ditto, why not malloc?
> 
> _[LibOS/shim/src/fs/proc/thread.c, line 632 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBI2g-2x8GH257ZoUt:-MUBI2g0802mSJtLm7_7:b-l0o1r3) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L632)):_
> 
> > ```c
> >     // Don't really need to do any work here, but keeping as a placeholder,
> >     // just in case.
> > ```
> 
> I don't understand this comment, what placeholder?
> 
> _[LibOS/shim/src/fs/proc/thread.c, line 647 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBJVo7CBaON18kM7yL:-MUBJVo7CBaON18kM7yM:b-d4ui68) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L647)):_
> 
> > ```c
> >     // Only used by one file
> >     __UNUSED(name);
> >     memset(buf, 0, sizeof(struct stat));
> > ```
> 
> `sizeof(struct stat)` -> `sizeof(*buf)`

Done. (some of these are not addressed, but as I mentioned, they are aligned with what we have in the code today.

dimakuv reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: all files reviewed, 6 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @mkow)



<!-- Sent from Reviewable.io -->

mkow reviewed (changes requested):
> Reviewed 1 of 2 files at r3, 2 of 3 files at r5, 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @dimakuv and @svenkata9)

___
*[LibOS/shim/src/fs/proc/thread.c, line 618 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBJIo32jNDO5MR45TB:-MVfnDfB0Ps15DStmJdk:b-9ejgys) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L618)):*
<details><summary><i>Previously, svenkata9 wrote…</i></summary><blockquote>

Done
</blockquote></details>

That's just replacing `calloc` with `malloc`, nothing more (I mean, this fix is trivial). We can fix then in one sweep in the old code, but why introduce the same problems in the new code?
___
*[LibOS/shim/src/fs/proc/thread.c, line 647 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBJVo7CBaON18kM7yL:-MVfmKu05fM-TzXIZXIi:befq9qc) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L647)):*
<details><summary><i>Previously, svenkata9 wrote…</i></summary><blockquote>

Done. (some of these are not addressed, but as I mentioned, they are aligned with what we have in the code today.
</blockquote></details>

We have a lot of legacy code written in ugly and bug-prone ways and this is not a good argument to justify writing new code the same way.


<!-- Sent from Reviewable.io -->

dimakuv reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @mkow)

___
*[LibOS/shim/src/fs/proc/thread.c, line 618 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBJIo32jNDO5MR45TB:-MWYWOSABN9iFoO-po4l:b-vnl330) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L618)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

That's just replacing `calloc` with `malloc`, nothing more (I mean, this fix is trivial). We can fix then in one sweep in the old code, but why introduce the same problems in the new code?
</blockquote></details>

Let's finish this PR and resolve this as per Michal's suggestion: replacing `calloc` with `malloc`.
___
*[LibOS/shim/src/fs/proc/thread.c, line 647 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBJVo7CBaON18kM7yL:-MWYWZcG-w6HBfuHM93s:bsu6ab) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L647)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

We have a lot of legacy code written in ugly and bug-prone ways and this is not a good argument to justify writing new code the same way.
</blockquote></details>

Let's finish this PR and resolve this as per Michal's suggestion: replacing `sizeof(struct stat)` -> `sizeof(*buf)`.


<!-- Sent from Reviewable.io -->

svenkata9 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: 3 of 4 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv and @mkow)

___
*[LibOS/shim/src/fs/proc/thread.c, line 618 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBJIo32jNDO5MR45TB:-MWxa7TBE8ApwWvSxPZ2:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L618)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Let's finish this PR and resolve this as per Michal's suggestion: replacing `calloc` with `malloc`.
</blockquote></details>

Done.
___
*[LibOS/shim/src/fs/proc/thread.c, line 647 at r2](https://reviewable.io/reviews/oscarlab/graphene/2180#-MUBJVo7CBaON18kM7yL:-MWxa5zhDgJlWfPZeNrP:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/0074b859ca250af546ce6a983c7ebf2852728eef/LibOS/shim/src/fs/proc/thread.c#L647)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Let's finish this PR and resolve this as per Michal's suggestion: replacing `sizeof(struct stat)` -> `sizeof(*buf)`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mkow reviewed (approved):
> Reviewed 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL)



<!-- Sent from Reviewable.io -->

dimakuv reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

dimakuv reviewed (dismissed):
> Reviewed 2 of 2 files at r8.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel)



<!-- Sent from Reviewable.io -->

dimakuv commented:
Jenkins, test this please

dimakuv reviewed (approved):
> Reviewed 1 of 1 files at r9.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel)



<!-- Sent from Reviewable.io -->

mkow reviewed (approved):
> Reviewed 2 of 2 files at r8, 1 of 1 files at r9.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

mkow commented:
Jenkins, retest this please

dimakuv reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2180)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->
